src/data/finsupp/lex.lean

```diff
@@ -0,0 +1,90 @@
+-/
+
+variables {α N : Type*}
+
+namespace finsupp
+
+lemma filter_ne_eq_empty_iff [decidable_eq α] [decidable_eq N] [has_zero N] {f g : α →₀ N} :
+  (f.support ∪ g.support).filter (λ a, f a ≠ g a) = ∅ ↔ f = g :=
+begin
+  refine ⟨λ h, _, λ h, h ▸ by simp⟩,
+  ext a,
+  contrapose h,
+  refine finset.ne_empty_of_mem (_ : a ∈ _),
+  simp only [ne.def, finset.mem_filter, finset.mem_union, finsupp.mem_support_iff],
+  exact ⟨not_and_distrib.1 $ mt (and_imp.2 eq.substr) (ne.symm h), h⟩,
+end
+
+section N_has_zero
+variables [has_zero N]
+
+/-- The lexicographic relation on `α →₀ N`, where `α` is ordered by `r`,
+  and `N` is ordered by `s`. -/
+protected def lex (r : α → α → Prop) (s : N → N → Prop) (x y : α →₀ N) : Prop :=
+pi.lex r (λ _, s) x y
+
+instance [has_lt α] [has_lt N] : has_lt (lex (α →₀ N)) := ⟨finsupp.lex (<) (<)⟩
+
+instance lex.is_strict_order [linear_order α] [partial_order N] :
+  is_strict_order (lex (α →₀ N)) (<) :=
+let i : is_strict_order (lex (α → N)) (<) := pi.lex.is_strict_order in
+{ irrefl := to_lex.surjective.forall.2 $ λ a, @irrefl _ _ i.to_is_irrefl a,
+  trans := to_lex.surjective.forall₃.2 $ λ a b c, @trans _ _ i.to_is_trans a b c }
+
+variables [linear_order α]
+
+/--  The partial order on `finsupp`s obtained by the lexicographic ordering.
+`finsupp.lex.linear_order` is the proof that this partial order is in fact linear. -/
+instance lex.partial_order [partial_order N] : partial_order (lex (α →₀ N)) :=
+partial_order.lift (λ x, to_lex ⇑(of_lex x)) finsupp.coe_fn_injective--fun_like.coe_injective
+
+variable [linear_order N]
+/--  The linear order on `finsupp`s obtained by the lexicographic ordering. -/
+noncomputable instance lex.linear_order : linear_order (lex (α →₀ N)) :=
+{ le_total := to_lex.surjective.forall₂.2 $ λ f g, begin
+    let dfug : finset α := (f.support ∪ g.support).filter (λ a, of_lex f a ≠ of_lex g a),
+    cases dfug.eq_empty_or_nonempty,
+    { exact or.inl (finsupp.filter_ne_eq_empty_iff.mp h).le },
+    { cases le_or_lt (of_lex f (dfug.min' h)) (of_lex g (dfug.min' h)) with mf mg,
+      { refine or.inl (or.inr _),
+        rcases finset.mem_filter.mp (finset.min'_mem _ h) with ⟨-, h⟩,
+        refine ⟨_, λ j hj, _, lt_of_le_of_ne mf h⟩, clear h,
+        by_cases js : j ∈ f.support ∪ g.support,
+        { contrapose! hj,
+          exact finset.min'_le _ _ (finset.mem_filter.mpr ⟨js, hj⟩) },
+        { simp only [finset.mem_union, not_or_distrib, finsupp.mem_support_iff, not_not] at js,
+          simp only [js, of_lex_to_lex, pi.to_lex_apply] } },
+      { refine or.inr (or.inr ⟨_, λ j hj, _, mg⟩),
+        by_cases js : j ∈ f.support ∪ g.support,
+        { contrapose! hj,
+          exact finset.min'_le _ _ (finset.mem_filter.mpr ⟨js, hj.symm⟩) },
+        { simp only [finset.mem_union, not_or_distrib, finsupp.mem_support_iff, not_not] at js,
+          simp only [js, of_lex_to_lex, pi.to_lex_apply] } } }
+    end,
+  decidable_le := by { classical, apply_instance },
```

Comment:
Is this computably decidable by interating the support?

Reply:
I am not sure that I understand your suggestion: should I try to prove the instance by induction on the support?

Reply:
As a first step, can you prove `decidable_rel (finsupp.lex r s)`?

Reply:
I would imagine the proof looks something like first showing
```
lemma lex_if : finsupp.lex r s f g ↔
∃ (i ∈ x.support), (∀ (j ∈ x.support), r j i → x j = y j) ∧ s (x i) (y i)
```
that is, adjusting the binders to be obviously finite. I don't think it's correct as written above, you probably need some disjunctions

Reply:
Hm, do you want to take a look at `wit` from the newly pushed #16026?  If I understand correctly, you seem to be wanting a decidable version of `wit`, correct?

Reply:
`wit` is defined [here](https://github.com/leanprover-community/mathlib/blob/d7977f9d133160d908e1f62256e0b038213719db/src/data/finsupp/witnesses.lean#L89).

Reply:
I won't have time to look at this soon, but basically the main input should be that you can decide the relative order of two finsupps by looking only at the union of their supports, filtering out the values where they agree, search for the minimum and evaluating the finsupps at this minimum (they are equal if nothing survives the filtering process).  Does this sound "decidable"?

Reply:
@eric-wieser, is what you are suggesting to prove [`le_iff_of_lex_apply_wit_le`](https://github.com/leanprover-community/mathlib/blob/d7977f9d133160d908e1f62256e0b038213719db/src/data/finsupp/witnesses.lean#L185)?

For context, the lemma says:
```lean
lemma le_iff_of_lex_apply_wit_le [nonempty α] [linear_order α] [linear_order N] [has_zero N]
  {f g : lex (α →₀ N)} :
  f ≤ g ↔ of_lex f (f.wit g) ≤ of_lex g (f.wit g) :=
```
where the definition of `wit` is
```lean
noncomputable def wit [has_zero N] (f g : α →₀ N) : α :=
dite (f.diff g).nonempty (λ h, (f.diff g).min' h) (λ _, nonempty.some ‹_›)
```
and `finsupp.diff` is
```lean
noncomputable def wit [has_zero N] (f g : α →₀ N) : α :=
dite (f.diff g).nonempty (λ h, (f.diff g).min' h) (λ _, nonempty.some ‹_›)
```

I am guessing that the non-computability could be replaced by providing an explicit element of `α` and removing `nonempty.some`.

Reply:
@eric-wieser, I could probably adapt the code in the other PR to work with arbitrary relations `r s` instead of `≠ <`.  However, I'm planning to use this order in the context of `add_monoid_algebra`s, where I do not think that decidability reach anyway.

Would it make sense to leave this for a separate PR, in case someone wants decidability?  I could leave a note in the file saying that decidability could be achieved (something I do not really know, but seems likely).

Reply:
Yes, I think leaving a note here is fine; this isn't quite as easy as I first thought.

Reply:
I left a "decidable" note.

Reply:
Presumably the dependency dissappears if we postpone the `decidable` instance?

Reply:
I'm sorry, I do not understand this comment: which dependency are you referring to?  I'm happy to change stuff, but I do not know what!  😀

Reply:
You added a dependency on another PR (the witness one)

Reply:
Oh, I see.  On lines 51 and following, I use `diff`.  I had that inlined with the definition, avoiding the dependency earlier.  However I suspect that `wit` will play a bigger role in the development of this file.  For this reason, I think that even removing now the dependency would be short-lived.

If you think that it is better for these two PR to not depend on one another, I can sever the ties temporarily, though.